src/include/codegen/hash_table.h

```diff
@@ -119,12 +130,23 @@ class HashTable {
                              ProbeCallback &probe_callback,
                              InsertCallback &insert_callback) const = 0;
 
+  // Probe the hash table and insert a new slot if needed, returning both the
+  // result and the data pointer
+  virtual ProbeResult ProbeOrInsert(
+      CodeGen &codegen, llvm::Value *ht_ptr, llvm::Value *hash,
+      const std::vector<codegen::Value> &key) const = 0;
+
   // Insert a new entry into the hash table with the given keys, but don't
   // perform any key matching or merging
   virtual void Insert(CodeGen &codegen, llvm::Value *ht_ptr, llvm::Value *hash,
                       const std::vector<codegen::Value> &keys,
                       InsertCallback &callback) const = 0;
 
+  // Insert a new entry into the hash table with the given keys, not caring
+  // about existing entries.
+  virtual void Insert(CodeGen &codegen, llvm::Value *ht_ptr, llvm::Value *hash,
+                      const std::vector<codegen::Value> &key) const = 0;
```

Comment:
What is the difference between the two `Insert` methods? Can you implement the one you created by using the existing `Insert` and providing a no-op callback? I ask because I don't want to complicate the interface too much.

Reply:
In the subclass both call the same method. It exists, because the `InsertCallback &insert_callback` reference can't be null. 
The three possibilities I see are: 
1. Changing signature to take pointer instead of reference (can be null)
2. Overload function (what I did)
3. Create dummy InsertCallback instance. Then however the functions are actually called during the lookup, even if they are empty (if not removed by compiler).

Reply:
If no extra work is done during insertions when a value is explicitly provided and when it is not, I'd prefer to have just one `Insert` method. Users not caring to insert values provide an `InsertCallback` that is a no-op. You can even create a `NoOnInsertCallback` implementation class in `codegen::HashTable` that is a no-op, which users of the hash table can use.